File: lib/utils/autoupdater.py
from .. import is_newer, __app_name__
from .application import application
from .constants import autoupdate_path, user_update_folder
from .logger import Logger
from threading import Thread

import json
import os
import requests
import shutil
import tempfile
import time


class Asset:
    def __init__(self, url, size):
        self.url = url
        self.size = size

    def __repr__(self):
        return f"Asset(url={self.url}, size={self.size})"


class AutoUpdater:
    def __init__(self):
        self.logger = Logger(self.__class__.__name__)
        self.app_image = os.getenv("APPIMAGE")

        if self.app_image is not None:
            self.logger.info(f"AppImage location: {self.app_image}")
            with open(autoupdate_path, "r") as file:
                data = json.load(file)
            self.owner = data["owner"]
            self.repository = data["repository"]
            self.update_path = None

            self.logger.info(
                f"AutoUpdater configured for repository {self.owner}/{self.repository}"
            )
        else:
            self.logger.warning("Auto update is only available for AppImage version")

    def start(self):
        if self.app_image is not None:
            Thread(name="AutoUpdater", target=self._check_task).start()

    def _check_task(self):
        time.sleep(1)
        while self.update_path is None:
            self.logger.info("Checking for updates...")
            data = self.get_update_url()
            if data is None:
                self.logger.info("No update found")
                time.sleep(3600)
            else:
                self.download_update(data.url, data.size)
                self.copy_file(self.update_path)
                application.relaunch_application()

    def copy_file(self, tmp_file):
        try:
            shutil.copy2(
                tmp_file, os.path.join(user_update_folder, f"{__app_name__}.AppImage")
            )
        except Exception as e:
            self.logger.error(f"Error while copying file: {e}")

    def get_update_url(self) -> Asset | None:
        url = f"https://api.github.com/repos/{self.owner}/{self.repository}/releases/latest"
        response = requests.get(url)

        if response.status_code == 200:
            release_data = response.json()
            remote_version = release_data["tag_name"]
            if is_newer(remote_version):
                self.logger.info(f"Update found for version {remote_version}")
                data = [
                    Asset(asset["browser_download_url"], asset["size"])
                    for asset in release_data["assets"]
                    if asset["name"].endswith(".AppImage")
                ]
                if len(data) > 0:
                    return data[0]
        else:
            print(f"Error getting latest release: {response.status_code}")

        return None

    def download_update(self, url, size):
        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
            temp_file_path = temp_file.name
            self.logger.info(f"Downloading {size} bytes from {url} to {temp_file_path}")

            try:
                start_time = time.time()
                response = requests.get(url, stream=True)
                response.raise_for_status()
                total_downloaded = 0

                with open(temp_file_path, "wb") as f:
                    for chunk in response.iter_content(chunk_size=8192):
                        f.write(chunk)
                        total_downloaded += len(chunk)

                end_time = time.time()
                download_time = end_time - start_time
                speed = total_downloaded / download_time

                speed_mb_per_s = speed / (1024 * 1024)
                self.logger.info(
                    f"Download completed in {download_time:.2f} seconds ({speed_mb_per_s:.2f} MB/s)"
                )

                self.update_path = temp_file_path

            except requests.exceptions.RequestException as e:
                self.logger.error(f"Error downloading the file: {e}")


auto_updater = AutoUpdater()

File: RogPerfTuner/src/utils/autoupdater.cpp
#ifndef IS_AURPKG

#include "../../include/utils/autoupdater.hpp"

#include <fstream>
#include <functional>
#include <iostream>
#include <string>
#include <thread>

#include "../../include/utils/file_utils.hpp"
#include "../../include/utils/time_utils.hpp"

#ifndef CPPHTTPLIB_OPENSSL_SUPPORT
#define CPPHTTPLIB_OPENSSL_SUPPORT
#endif

#include <nlohmann/json.hpp>

#include "httplib.h"

using json = nlohmann::json;

bool AutoUpdater::is_newer(const std::string& remote_version) const {
	auto parse_version = [](const std::string& v) {
		std::vector<int> parts;
		size_t start = 0;
		size_t end	 = v.find('.');
		while (end != std::string::npos) {
			parts.push_back(std::stoi(v.substr(start, end - start)));
			start = end + 1;
			end	  = v.find('.', start);
		}
		parts.push_back(std::stoi(v.substr(start)));
		return parts;
	};

	auto v1 = parse_version(Constants::APP_VERSION);
	auto v2 = parse_version(remote_version);

	return v1 < v2;
}

Asset AutoUpdater::get_update_url() {
	std::string url = "/repos/" + owner + "/" + repository + "/releases/latest";
	httplib::SSLClient cli("api.github.com", 443);
	cli.set_default_headers({{"User-Agent", "AutoUpdater"}});
	auto res = cli.Get(url.c_str());

	if (!res || res->status != 200) {
		logger.error("Error getting latest release: {}", res ? res->status : 0);
		return Asset("", 0);
	}

	try {
		auto release_data		   = json::parse(res->body);
		std::string remote_version = release_data["tag_name"];

		if (!is_newer(remote_version)) {
			return Asset("", 0);
		}

		// Buscar assets que terminen con ".AppImage"
		for (const auto& asset : release_data["assets"]) {
			std::string name = asset["name"];
			if (name.size() >= 9 && name.substr(name.size() - 9) == ".AppImage") {
				std::string url = asset["browser_download_url"];
				size_t size		= asset["size"];
				logger.info("Update found: {}({} bytes)", name, std::to_string(size));
				return Asset(url, size);
			}
		}

		logger.info("No AppImage asset found in latest release");
		return Asset("", 0);
	} catch (const std::exception& e) {
		logger.error("JSON parse error: {}", e.what());
		return Asset("", 0);
	}
}

void AutoUpdater::download_update(const Asset& asset) {
	logger.info("Downloading update...");

	// Separar host y path de la URL
	std::regex url_regex(R"(https://([^/]+)(/.+))");
	std::smatch match;
	if (!std::regex_match(asset.url, match, url_regex)) {
		logger.error("Invalid URL: {}", asset.url);
		return;
	}

	std::string host = match[1];
	std::string path = match[2];

	httplib::SSLClient cli(host.c_str(), 443);
	cli.set_follow_location(true);
	cli.set_default_headers({{"User-Agent", "AutoUpdater"}});

	auto res = cli.Get(path.c_str());
	if (!res || res->status != 200) {
		logger.error("Failed to download file, HTTP status: {}", res ? res->status : 0);
		return;
	}

	if (res && res->status == 200) {
		std::ofstream ofs(Constants::UPDATE_TMP_FILE, std::ios::binary);
		ofs << res->body;
		ofs.close();
		logger.info("Download completed");
	} else {
		logger.error("Error downloading the file");
	}
}

void AutoUpdater::copy_file() {
	try {
		FileUtils::move(Constants::UPDATE_TMP_FILE, Constants::UPDATE_FILE);
		logger.info("File copied to update folder");
	} catch (const std::exception& e) {
		logger.error("Error copying file: {}", e.what());
	}
}

void AutoUpdater::check_task() {
	if (FileUtils::exists(Constants::UPDATE_TMP_FILE)) {
		FileUtils::remove(Constants::UPDATE_TMP_FILE);
	}

	TimeUtils::sleep(1000);

	while (!FileUtils::exists(Constants::UPDATE_TMP_FILE)) {
		logger.info("Checking for updates...");
		Asset asset = get_update_url();
		if (asset.url.empty()) {
			logger.info("No update found");
			TimeUtils::sleep(CHECK_INTERVAL);
		} else {
			logger.info("Update found");
			while (perform_update_check && !perform_update_check()) {
				logger.info("Update blocked by application, retry in 10 minutes");
				TimeUtils::sleep(10 * 60 * 1000);
			}

			download_update(asset);
			if (!Constants::DEV_MODE) {
				copy_file();
				if (restart_method) {
					restart_method();
				}
			}
		}
	}
}

AutoUpdater::AutoUpdater(Callback restart_method_, std::function<bool()> perform_update_check_)
	: Loggable("AutoUpdater"), restart_method(restart_method_), perform_update_check(perform_update_check_) {
	std::thread(&AutoUpdater::check_task, this).detach();
}
#endif